logp/logger.go

```diff
@@ -50,6 +52,32 @@ func NewLogger(selector string, options ...LogOption) *Logger {
 	return newLogger(loadLogger().rootLogger, selector, options...)
 }
 
+// NewInMemory returns a new in-memory logger along with the buffer to which i
+// logs.
+// encCfg configures the log format, use logp.ConsoleEncoderConfig for console
+// format, logp.JSONEncoderConfig for JSON or any other valid zapcore.EncoderConfig.
+func NewInMemory(selector string, encCfg zapcore.EncoderConfig) (*Logger, *bytes.Buffer) {
```

Comment:
question on returning the pointer to the bytes.Buffer.  It looks like zap uses it as a WriteSyncer, which does have a `Lock` function to make it safe for concurrent use.  If someone has the pointer to the underlying Buffer they could get around that.  Also is it a problem if someone does something like `Reset` on the Buffer?

Reply:
> It looks like zap uses it as a WriteSyncer, which does have a Lock function to make it safe for concurrent use.

Almost, it's a bit confusing, it has a `writerWrapper` which just adds a no-op `Sync()` to an `io.Writer` and also the `Lock` function, this adds a lock to a `WriteSyncer`, the interface, making it goroutinesafe.

I added the lock so it's safe to be used anywhere and an improved the docs